File: scripts/overnight_web_scrape.mjs
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { execSync } from 'child_process';
import Database from 'better-sqlite3';

const WORKSPACE = '/Users/corgi12/.eragon-joshua_augustine/joshua_augustine_workspace';
const DB_PATH = `${WORKSPACE}/jordan.ai/pipeline/master.db`;
const CHECKPOINT = `${WORKSPACE}/jordan.ai/pipeline/overnight_scrape_checkpoint.json`;
const OUTPUT = `${WORKSPACE}/jordan.ai/pipeline/overnight_scrape_results.json`;
const CONCURRENCY = 5;
const DELAY_MS = 500;
const TIMEOUT_MS = 8000;

// Load checkpoint
let checkpoint = { processed: 0, results: {} };
if (existsSync(CHECKPOINT)) {
  checkpoint = JSON.parse(readFileSync(CHECKPOINT, 'utf8'));
  console.log(`Resuming from checkpoint: ${checkpoint.processed} processed, ${Object.keys(checkpoint.results).length} results`);
}

// Get domains to scrape — qualified companies without enrichment signals
const db = new Database(DB_PATH, { readonly: true });
const rows = db.prepare(`
  SELECT domain, company_name, vibe_score, blueprint_score, grade
  FROM companies 
  WHERE domain IS NOT NULL AND domain != ''
  AND (dq_reason IS NULL OR dq_reason = '')
  AND vibe_score IS NOT NULL AND vibe_score > 0
  AND blueprint_score IS NOT NULL AND blueprint_score >= 50
  AND (hiring_signal IS NULL OR hiring_signal = '' OR hiring_signal = '0')
  AND (recent_news IS NULL OR recent_news = '')
  ORDER BY blueprint_score DESC
`).all();
db.close();

console.log(`Companies to scrape: ${rows.length}`);

async function fetchWithTimeout(url, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const r = await fetch(url, { 
      signal: controller.signal,
      headers: { 'User-Agent': 'Mozilla/5.0 (compatible; research-bot)' },
      redirect: 'follow'
    });
    clearTimeout(timer);
    if (!r.ok) return null;
    const text = await r.text();
    return text.substring(0, 50000); // cap at 50KB
  } catch {
    clearTimeout(timer);
    return null;
  }
}

function extractSignals(html, domain) {
  if (!html) return {};
  const lower = html.toLowerCase();
  const signals = {};
  
  // Hiring signals
  const jobKeywords = ['careers', 'we\'re hiring', 'join our team', 'open positions', 'job openings', 'work with us', 'apply now'];
  const hiringMatches = jobKeywords.filter(k => lower.includes(k));
  if (hiringMatches.length > 0) {
    signals.hiring = true;
    signals.hiring_keywords = hiringMatches;
  }
  
  // GitHub link
  const ghMatch = html.match(/github\.com\/([a-zA-Z0-9_-]+)/);
  if (ghMatch && !['topics', 'features', 'pricing', 'about', 'login', 'signup', 'join'].includes(ghMatch[1].toLowerCase())) {
    signals.github = `https://github.com/${ghMatch[1]}`;
  }
  
  // Twitter/X
  const twMatch = html.match(/(?:twitter\.com|x\.com)\/([a-zA-Z0-9_]+)/);
  if (twMatch && !['share', 'intent', 'home', 'login', 'signup'].includes(twMatch[1].toLowerCase())) {
    signals.twitter = twMatch[1];
  }
  
  // Tech stack signals from meta/scripts
  const techSignals = [];
  if (lower.includes('react') || lower.includes('next.js') || lower.includes('nextjs')) techSignals.push('React');
  if (lower.includes('vue') || lower.includes('nuxt')) techSignals.push('Vue');
  if (lower.includes('angular')) techSignals.push('Angular');
  if (lower.includes('stripe')) techSignals.push('Stripe');
  if (lower.includes('intercom')) techSignals.push('Intercom');
  if (lower.includes('hubspot')) techSignals.push('HubSpot');
  if (lower.includes('segment')) techSignals.push('Segment');
  if (lower.includes('mixpanel')) techSignals.push('Mixpanel');
  if (lower.includes('amplitude')) techSignals.push('Amplitude');
  if (lower.includes('sentry')) techSignals.push('Sentry');
  if (lower.includes('datadog')) techSignals.push('Datadog');
  if (lower.includes('aws') || lower.includes('amazon web services')) techSignals.push('AWS');
  if (lower.includes('google cloud') || lower.includes('gcp')) techSignals.push('GCP');
  if (lower.includes('azure')) techSignals.push('Azure');
  if (lower.includes('kubernetes') || lower.includes('k8s')) techSignals.push('K8s');
  if (lower.includes('docker')) techSignals.push('Docker');
  if (lower.includes('tensorflow') || lower.includes('pytorch') || lower.includes('openai')) techSignals.push('AI/ML');
  if (techSignals.length > 0) signals.tech_stack = techSignals;
  
  // Team size hints
  const teamMatch = lower.match(/(\d+)\+?\s*(?:team members|employees|people|engineers)/);
  if (teamMatch) {
    signals.team_size_hint = parseInt(teamMatch[1]);
  }
  
  // Funding mentions
  const fundingPatterns = [/raised\s*\$[\d.]+[mkb]/i, /series\s*[a-f]/i, /seed\s*(?:round|funding)/i, /\$[\d.]+[mkb]\s*(?:funding|round|raise)/i];
  const fundingHits = fundingPatterns.filter(p => p.test(html));
  if (fundingHits.length > 0) {
    signals.has_funding_mention = true;
  }
  
  // Blog/content signals
  if (lower.includes('/blog') || lower.includes('blog.')) signals.has_blog = true;
  
  // Pricing page = real product
  if (lower.includes('/pricing') || lower.includes('pricing')) signals.has_pricing = true;
  
  return signals;
}

async function scrapeBatch(batch) {
  const results = [];
  for (const row of batch) {
    if (checkpoint.results[row.domain]) continue;
    
    const url = `https://${row.domain}`;
    const html = await fetchWithTimeout(url, TIMEOUT_MS);
    const signals = extractSignals(html, row.domain);
    signals.scraped = true;
    signals.has_content = !!html;
    
    checkpoint.results[row.domain] = signals;
    results.push({ domain: row.domain, ...signals });
    
    await new Promise(r => setTimeout(r, DELAY_MS));
  }
  return results;
}

async function run() {
  const startIdx = checkpoint.processed;
  const remaining = rows.slice(startIdx);
  console.log(`Starting from index ${startIdx}, ${remaining.length} remaining`);
  
  for (let i = 0; i < remaining.length; i++) {
    const row = remaining[i];
    if (checkpoint.results[row.domain]) {
      checkpoint.processed = startIdx + i + 1;
      continue;
    }
    
    const url = `https://${row.domain}`;
    const html = await fetchWithTimeout(url, TIMEOUT_MS);
    const signals = extractSignals(html, row.domain);
    signals.scraped = true;
    signals.has_content = !!html;
    checkpoint.results[row.domain] = signals;
    checkpoint.processed = startIdx + i + 1;
    
    if ((startIdx + i + 1) % 500 === 0) {
      writeFileSync(CHECKPOINT, JSON.stringify({ processed: checkpoint.processed, results: checkpoint.results }));
      
      const total = Object.keys(checkpoint.results).length;
      const hiring = Object.values(checkpoint.results).filter(r => r.hiring).length;
      const github = Object.values(checkpoint.results).filter(r => r.github).length;
      const funding = Object.values(checkpoint.results).filter(r => r.has_funding_mention).length;
      const tech = Object.values(checkpoint.results).filter(r => r.tech_stack?.length > 0).length;
      console.log(`${checkpoint.processed}/${rows.length} | hiring:${hiring} github:${github} funding:${funding} tech:${tech}`);
    }
    
    await new Promise(r => setTimeout(r, DELAY_MS));
  }
  
  // Final save
  writeFileSync(CHECKPOINT, JSON.stringify({ processed: checkpoint.processed, results: checkpoint.results }));
  writeFileSync(OUTPUT, JSON.stringify(checkpoint.results, null, 2));
  
  const total = Object.keys(checkpoint.results).length;
  const hiring = Object.values(checkpoint.results).filter(r => r.hiring).length;
  const github = Object.values(checkpoint.results).filter(r => r.github).length;
  const funding = Object.values(checkpoint.results).filter(r => r.has_funding_mention).length;
  const tech = Object.values(checkpoint.results).filter(r => r.tech_stack?.length > 0).length;
  const pricing = Object.values(checkpoint.results).filter(r => r.has_pricing).length;
  const blog = Object.values(checkpoint.results).filter(r => r.has_blog).length;
  
  console.log(`\n=== DONE ===`);
  console.log(`Scraped: ${total}`);
  console.log(`Hiring signals: ${hiring}`);
  console.log(`GitHub links: ${github}`);
  console.log(`Funding mentions: ${funding}`);
  console.log(`Tech stack detected: ${tech}`);
  console.log(`Has pricing page: ${pricing}`);
  console.log(`Has blog: ${blog}`);
}

run().catch(e => console.error('Fatal:', e));

File: scripts/overnight_web_scrape.py
#!/usr/bin/env python3
"""Overnight website scraper — extracts hiring, GitHub, tech stack, funding signals from homepages.
Uses only free HTTP requests. Checkpoints every 500 companies."""

import sqlite3, json, re, os, time, sys
from urllib.request import urlopen, Request
from urllib.error import URLError, HTTPError
from concurrent.futures import ThreadPoolExecutor, as_completed

WORKSPACE = '/Users/corgi12/.eragon-joshua_augustine/joshua_augustine_workspace'
DB_PATH = f'{WORKSPACE}/jordan.ai/pipeline/master.db'
CHECKPOINT = f'{WORKSPACE}/jordan.ai/pipeline/overnight_scrape_checkpoint.json'
OUTPUT = f'{WORKSPACE}/jordan.ai/pipeline/overnight_scrape_results.json'
TIMEOUT = 5
DELAY = 0.2

# Load checkpoint
checkpoint = {'processed': 0, 'results': {}}
if os.path.exists(CHECKPOINT):
    with open(CHECKPOINT) as f:
        checkpoint = json.load(f)
    print(f"Resuming: {checkpoint['processed']} processed, {len(checkpoint['results'])} results")

# Get domains to scrape
db = sqlite3.connect(DB_PATH)
db.row_factory = sqlite3.Row
rows = db.execute("""
    SELECT domain, company_name, vibe_score, blueprint_score, grade
    FROM companies 
    WHERE domain IS NOT NULL AND domain != ''
    AND (dq_reason IS NULL OR dq_reason = '')
    AND vibe_score IS NOT NULL AND vibe_score > 0
    AND blueprint_score IS NOT NULL AND blueprint_score >= 50
    ORDER BY blueprint_score DESC
""").fetchall()
db.close()

print(f"Qualified companies to scrape: {len(rows)}")

def fetch_page(domain):
    url = f'https://{domain}'
    req = Request(url, headers={'User-Agent': 'Mozilla/5.0 (compatible; research-bot/1.0)'})
    try:
        with urlopen(req, timeout=TIMEOUT) as r:
            return r.read(50000).decode('utf-8', errors='ignore')
    except:
        try:
            url2 = f'http://{domain}'
            req2 = Request(url2, headers={'User-Agent': 'Mozilla/5.0 (compatible; research-bot/1.0)'})
            with urlopen(req2, timeout=TIMEOUT) as r:
                return r.read(50000).decode('utf-8', errors='ignore')
        except:
            return None

def extract_signals(html, domain):
    if not html:
        return {'scraped': True, 'has_content': False}
    
    lower = html.lower()
    signals = {'scraped': True, 'has_content': True}
    
    # Emails — extract personal emails from page (skip generic)
    generic_prefixes = {'info','sales','support','contact','hello','admin','help','office','team',
                        'marketing','billing','press','media','legal','hr','jobs','careers','noreply',
                        'no-reply','webmaster','postmaster','abuse','privacy','security','compliance'}
    email_pattern = re.findall(r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}', html)
    personal_emails = []
    seen = set()
    for e in email_pattern:
        e_lower = e.lower()
        prefix = e_lower.split('@')[0]
        e_domain = e_lower.split('@')[1] if '@' in e_lower else ''
        # Skip generic prefixes, image files, CSS, example domains
        if prefix in generic_prefixes:
            continue
        if any(e_lower.endswith(x) for x in ['.png','.jpg','.gif','.svg','.css','.js']):
            continue
        if 'example.com' in e_lower or 'sentry.io' in e_lower or 'email.com' in e_lower:
            continue
        if e_lower not in seen:
            seen.add(e_lower)
            personal_emails.append(e)
    if personal_emails:
        signals['emails'] = personal_emails[:5]  # cap at 5
    
    # Also try /contact and /about pages for more emails
    # (done in a second pass below)
    
    # Hiring
    job_kw = ["careers", "we're hiring", "join our team", "open positions", 
              "job openings", "work with us", "apply now", "view openings"]
    hits = [k for k in job_kw if k in lower]
    if hits:
        signals['hiring'] = True
        signals['hiring_count'] = len(hits)
    
    # GitHub
    gh = re.search(r'github\.com/([a-zA-Z0-9_-]+)', html)
    if gh and gh.group(1).lower() not in ('topics','features','pricing','about','login','signup','join','search'):
        signals['github'] = f'https://github.com/{gh.group(1)}'
    
    # Twitter/X
    tw = re.search(r'(?:twitter\.com|x\.com)/([a-zA-Z0-9_]+)', html)
    if tw and tw.group(1).lower() not in ('share','intent','home','login','signup','i'):
        signals['twitter'] = tw.group(1)
    
    # Tech stack
    tech = []
    checks = [
        ('React', ['react', 'next.js', 'nextjs', '__next']),
        ('Vue', ['vue', 'nuxt']),
        ('Angular', ['angular']),
        ('Stripe', ['stripe']),
        ('Intercom', ['intercom']),
        ('HubSpot', ['hubspot']),
        ('Segment', ['segment.com', 'analytics.js']),
        ('Mixpanel', ['mixpanel']),
        ('Amplitude', ['amplitude']),
        ('Sentry', ['sentry']),
        ('Datadog', ['datadog']),
        ('AWS', ['amazonaws.com']),
        ('GCP', ['googleapis.com', 'google cloud']),
        ('Cloudflare', ['cloudflare']),
        ('AI/ML', ['tensorflow', 'pytorch', 'openai', 'machine learning', 'artificial intelligence']),
        ('Vercel', ['vercel']),
        ('Netlify', ['netlify']),
    ]
    for name, keywords in checks:
        if any(k in lower for k in keywords):
            tech.append(name)
    if tech:
        signals['tech_stack'] = tech
    
    # Team size hint
    team = re.search(r'(\d+)\+?\s*(?:team members|employees|people|engineers|staff)', lower)
    if team:
        signals['team_size_hint'] = int(team.group(1))
    
    # Funding
    if re.search(r'(?:raised|funding|series\s*[a-f]|seed\s*round|\$[\d.]+[mkb]\s*(?:funding|round))', lower):
        signals['has_funding'] = True
    
    # Product signals
    if '/pricing' in lower or '"pricing"' in lower:
        signals['has_pricing'] = True
    if '/blog' in lower or 'blog.' in lower:
        signals['has_blog'] = True
    if '/docs' in lower or '/documentation' in lower or 'developer' in lower:
        signals['has_docs'] = True
    if '/api' in lower or 'api reference' in lower:
        signals['has_api'] = True
    
    return signals

# Main loop
start_idx = checkpoint['processed']
remaining = rows[start_idx:]
print(f"Starting from {start_idx}, {len(remaining)} to go")

for i, row in enumerate(remaining):
    domain = row['domain']
    
    if domain in checkpoint['results']:
        checkpoint['processed'] = start_idx + i + 1
        continue
    
    html = fetch_page(domain)
    signals = extract_signals(html, domain)
    
    # Second pass: if no emails found on homepage, try /contact and /about
    if not signals.get('emails'):
        for subpage in ['/contact', '/contact-us']:
            sub_html = fetch_page(f"{domain}{subpage}")
            if sub_html:
                sub_signals = extract_signals(sub_html, domain)
                if sub_signals.get('emails'):
                    signals['emails'] = sub_signals['emails']
                    signals['email_source'] = subpage
                    break
            time.sleep(0.2)
    
    checkpoint['results'][domain] = signals
    checkpoint['processed'] = start_idx + i + 1
    
    idx = start_idx + i + 1
    if idx % 100 == 0:
        with open(CHECKPOINT, 'w') as f:
            json.dump(checkpoint, f)
        
        r = checkpoint['results']
        total = len(r)
        hiring = sum(1 for v in r.values() if v.get('hiring'))
        github = sum(1 for v in r.values() if v.get('github'))
        funding = sum(1 for v in r.values() if v.get('has_funding'))
        tech = sum(1 for v in r.values() if v.get('tech_stack'))
        pricing = sum(1 for v in r.values() if v.get('has_pricing'))
        content = sum(1 for v in r.values() if v.get('has_content'))
        emails = sum(1 for v in r.values() if v.get('emails'))
        print(f"{idx}/{len(rows)} | content:{content} emails:{emails} hiring:{hiring} github:{github} funding:{funding} tech:{tech} pricing:{pricing}")
        sys.stdout.flush()
    
    time.sleep(DELAY)

# Final save
with open(CHECKPOINT, 'w') as f:
    json.dump(checkpoint, f)
with open(OUTPUT, 'w') as f:
    json.dump(checkpoint['results'], f, indent=2)

r = checkpoint['results']
print(f"\n=== DONE ===")
print(f"Scraped: {len(r)}")
print(f"Has content: {sum(1 for v in r.values() if v.get('has_content'))}")
print(f"Hiring signals: {sum(1 for v in r.values() if v.get('hiring'))}")
print(f"GitHub links: {sum(1 for v in r.values() if v.get('github'))}")
print(f"Funding mentions: {sum(1 for v in r.values() if v.get('has_funding'))}")
print(f"Tech stack: {sum(1 for v in r.values() if v.get('tech_stack'))}")
print(f"Has pricing: {sum(1 for v in r.values() if v.get('has_pricing'))}")
print(f"Has blog: {sum(1 for v in r.values() if v.get('has_blog'))}")
print(f"Has docs/API: {sum(1 for v in r.values() if v.get('has_docs') or v.get('has_api'))}")
emails_found = sum(1 for v in r.values() if v.get('emails'))
total_emails = sum(len(v.get('emails', [])) for v in r.values())
print(f"Companies with emails: {emails_found}")
print(f"Total email addresses: {total_emails}")
